podman4 can't build any code with quarkus builder

podman info
```
[root@~]# podman version
Client:       Podman Engine
Version:      4.3.1
API Version:  4.3.1
Go Version:   go1.18.7
Built:        Fri Nov 11 10:24:13 2022
OS/Arch:      linux/amd64
```
build command incloud builder:
```
mvn -ntp -f /root/alarm-service/pom.xml -DskipTests clean package -Pnative -Dquarkus.container-image.build=false -Dquarkus.native.container-runtime=podman -Dquarkus.container-image.tag=v1.0.0_20230131 -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.3-java11
mvn -ntp -f /root/alarm-service/pom.xml -DskipTests clean package -Pnative -Dquarkus.container-image.build=false -Dquarkus.native.container-runtime=podman -Dquarkus.container-image.tag=v1.0.0_20230131 -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.3-java11
mvn -ntp -f /root/alarm-service/pom.xml -DskipTests clean package -Pnative -Dquarkus.container-image.build=false -Dquarkus.native.container-runtime=podman -Dquarkus.container-image.tag=v1.0.0_20230131 -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-graalvmce-builder-image:22.1-java11
mvn -ntp -f /root/alarm-service/pom.xml -DskipTests clean package -Pnative -Dquarkus.container-image.build=false -Dquarkus.native.container-runtime=podman -Dquarkus.container-image.tag=v1.0.0_20230131 -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.1-java11
```

error info:
```
[ERROR] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Unable to get GraalVM version from the native-image binary.
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildRunner] podman run --env LANG=C --rm --user 0:0 --userns=keep-id -v /root/alarm-service/target/alarm-service-1.0-SNAPSHOT-native-image-source-jar:/project:z --name build-native-PTYgJ quay.io/quarkus/ubi-quarkus-mandrel-builder-image:22.1-java11 -J-DCoordinatorEnvironmentBean.transactionStatusManagerEnable=false -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dlogging.initial-configurator.min-level=500 -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=3 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Duser.language=en -J-Duser.country=US -J-Dfile.encoding=UTF-8 --features=io.quarkus.runner.Feature,io.quarkus.jdbc.postgresql.runtime.graal.SQLXMLFeature,io.quarkus.runtime.graal.ResourcesFeature,io.quarkus.runtime.graal.DisableLoggingFeature -H:-ParseOnce -J--add-exports=java.security.jgss/sun.security.krb5=ALL-UNNAMED -J--add-opens=java.base/java.text=ALL-UNNAMED -J--add-opens=java.base/java.io=ALL-UNNAMED -J--add-opens=java.base/java.lang.invoke=ALL-UNNAMED -J--add-opens=java.base/java.util=ALL-UNNAMED -H:ResourceConfigurationFiles=resources-config.json -H:ReflectionConfigurationFiles=reflection-config.json -H:+AllowFoldMethods -J-Djava.awt.headless=true --no-fallback --link-at-build-time -H:+ReportExceptionStackTraces -H:-AddAllCharsets --enable-url-protocols=http,https -H:-UseServiceLoaderFeature -H:+StackTrace -J--add-exports=org.graalvm.sdk/org.graalvm.nativeimage.impl=ALL-UNNAMED -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk=ALL-UNNAMED -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.configure=ALL-UNNAMED -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.localization=ALL-UNNAMED --exclude-config io\.netty\.netty-codec /META-INF/native-image/io\.netty/netty-codec/generated/handlers/reflect-config\.json --exclude-config io\.netty\.netty-handler /META-INF/native-image/io\.netty/netty-handler/generated/handlers/reflect-config\.json alarm-service-1.0-SNAPSHOT-runner -jar alarm-service-1.0-SNAPSHOT-runner.jar
Error: keep-id is only supported in rootless mode
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  29.799 s
[INFO] Finished at: 2023-01-31T22:15:09-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.16.0.Final:build (default) on project alarm-service: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: io.quarkus.deployment.pkg.steps.NativeImageBuildStep$ImageGenerationFailureException: Image generation failed. Exit code: 125
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.imageGenerationFailed(NativeImageBuildStep.java:421)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:262)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:834)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

The weird thing is that I dropped podman4 to podman3 but it worked

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
@@ -31,7 +31,8 @@ public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContai
                 String gid = getLinuxID("-gr");
                 if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
                     Collections.addAll(containerRuntimeArgs, "--user", uid + ":" + gid);
-                    if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
+                    if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN
+                            && containerRuntime.isRootless()) {
                         // Needed to avoid AccessDeniedExceptions
                         containerRuntimeArgs.add("--userns=keep-id");
                     }

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildLocalContainerRunner.java
@@ -2,27 +2,16 @@ package io.quarkus.deployment.pkg.steps;
 
 import static io.quarkus.deployment.pkg.steps.LinuxIDUtil.getLinuxID;
 
-import java.io.File;
-import java.io.IOException;
-import java.nio.file.Files;
-import java.nio.file.LinkOption;
 import java.nio.file.Path;
-import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
-import java.util.Objects;
-import java.util.Set;
-import java.util.function.Predicate;
-import java.util.stream.Collectors;
 
 import org.apache.commons.lang3.SystemUtils;
 import org.jboss.logging.Logger;
 
-import io.quarkus.deployment.OutputFilter;
 import io.quarkus.deployment.pkg.NativeConfig;
-import io.quarkus.deployment.util.ExecUtil;
 import io.quarkus.deployment.util.FileUtil;
 import io.quarkus.runtime.util.ContainerRuntimeUtil;
 
@@ -34,7 +23,8 @@ public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContai
         super(nativeConfig, outputDir);
         if (SystemUtils.IS_OS_LINUX) {
             ArrayList<String> containerRuntimeArgs = new ArrayList<>(Arrays.asList(baseContainerRuntimeArgs));
-            if (isDockerRootless(containerRuntime)) {
+            if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.DOCKER
+                    && containerRuntime.isRootless()) {
                 Collections.addAll(containerRuntimeArgs, "--user", String.valueOf(0));
             } else {
                 String uid = getLinuxID("-ur");
@@ -49,63 +39,6 @@ public class NativeImageBuildLocalContainerRunner extends NativeImageBuildContai
             }
             baseContainerRuntimeArgs = containerRuntimeArgs.toArray(baseContainerRuntimeArgs);
         }
-    }
-
-    private static boolean isDockerRootless(ContainerRuntimeUtil.ContainerRuntime containerRuntime) {
-        if (containerRuntime != ContainerRuntimeUtil.ContainerRuntime.DOCKER) {
-            return false;
-        }
-        String dockerEndpoint = fetchDockerEndpoint();
-        // docker socket?
-        String socketUriPrefix = "unix://";
-        if (dockerEndpoint == null || !dockerEndpoint.startsWith(socketUriPrefix)) {
-            return false;
-        }
-        String dockerSocket = dockerEndpoint.substring(socketUriPrefix.length());
-        String currentUid = getLinuxID("-ur");
-        if (currentUid == null || currentUid.isEmpty() || currentUid.equals(String.valueOf(0))) {
-            return false;
-        }
-
-        int socketOwnerUid;
-        try {
-            socketOwnerUid = (int) Files.getAttribute(Path.of(dockerSocket), "unix:uid", LinkOption.NOFOLLOW_LINKS);
-        } catch (IOException e) {
-            LOGGER.infof("Owner UID lookup on '%s' failed with '%s'", dockerSocket, e.getMessage());
-            return false;
-        }
-        return currentUid.equals(String.valueOf(socketOwnerUid));
-    }
-
-    private static String fetchDockerEndpoint() {
-        // DOCKER_HOST environment variable overrides the active context
-        String dockerHost = System.getenv("DOCKER_HOST");
-        if (dockerHost != null) {
-            return dockerHost;
-        }
-
-        OutputFilter outputFilter = new OutputFilter();
-        if (!ExecUtil.execWithTimeout(new File("."), outputFilter, Duration.ofMillis(3000),
-                "docker", "context", "ls", "--format",
-                "{{- if .Current -}} {{- .DockerEndpoint -}} {{- end -}}")) {
-            LOGGER.debug("Docker context lookup didn't succeed in time");
-            return null;
-        }
-
-        Set<String> endpoints = outputFilter.getOutput().lines()
-                .filter(Objects::nonNull)
-                .filter(Predicate.not(String::isBlank))
-                .collect(Collectors.toSet());
-        if (endpoints.size() == 1) {
-            return endpoints.stream().findFirst().orElse(null);
-        }
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debugf("Found too many active Docker endpoints: [%s]",
-                    endpoints.stream()
-                            .map(endpoint -> String.format("'%s'", endpoint))
-                            .collect(Collectors.joining(",")));
-        }
-        return null;
     }
 
     @Override

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/UpxCompressionBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/UpxCompressionBuildStep.java
@@ -124,7 +124,8 @@ public class UpxCompressionBuildStep {
             String gid = getLinuxID("-gr");
             if (uid != null && gid != null && !uid.isEmpty() && !gid.isEmpty()) {
                 Collections.addAll(commandLine, "--user", uid + ":" + gid);
-                if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN) {
+                if (containerRuntime == ContainerRuntimeUtil.ContainerRuntime.PODMAN
+                        && containerRuntime.isRootless()) {
                     // Needed to avoid AccessDeniedExceptions
                     commandLine.add("--userns=keep-id");
                 }

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
@@ -2,6 +2,7 @@ package io.quarkus.runtime.util;
 
 import java.io.BufferedReader;
 import java.io.IOException;
+import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
 import java.util.function.Predicate;
@@ -68,16 +69,24 @@ public final class ContainerRuntimeUtil {
             ProcessBuilder pb = new ProcessBuilder(containerRuntime.getExecutableName(), "info")
                     .redirectErrorStream(true);
             rootlessProcess = pb.start();
-            rootlessProcess.waitFor();
-            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(rootlessProcess.getInputStream()));
-            Predicate<String> stringPredicate;
-            // Docker includes just "rootless" under SecurityOptions, while podman includes "rootless: <boolean>"
-            if (containerRuntime == ContainerRuntime.DOCKER) {
-                stringPredicate = line -> line.trim().equals("rootless");
-            } else {
-                stringPredicate = line -> line.trim().equals("rootless: true");
+            int exitCode = rootlessProcess.waitFor();
+            if (exitCode != 0) {
+                log.warnf("Command \"%s\" exited with error code %d. " +
+                        "Rootless container runtime detection might not be reliable.",
+                        containerRuntime.getExecutableName(), exitCode);
             }
-            return bufferedReader.lines().anyMatch(stringPredicate);
+            try (InputStream inputStream = rootlessProcess.getInputStream();
+                    InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
+                    BufferedReader bufferedReader = new BufferedReader(inputStreamReader)) {
+                Predicate<String> stringPredicate;
+                // Docker includes just "rootless" under SecurityOptions, while podman includes "rootless: <boolean>"
+                if (containerRuntime == ContainerRuntime.DOCKER) {
+                    stringPredicate = line -> line.trim().equals("rootless");
+                } else {
+                    stringPredicate = line -> line.trim().equals("rootless: true");
+                }
+                return bufferedReader.lines().anyMatch(stringPredicate);
+            }
         } catch (IOException | InterruptedException e) {
             // If an exception is thrown in the process, assume we are not running rootless (default docker installation)
             log.debugf(e, "Failure to read info output from %s", containerRuntime.getExecutableName());

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ContainerRuntimeUtil.java
@@ -1,7 +1,10 @@
 package io.quarkus.runtime.util;
 
+import java.io.BufferedReader;
 import java.io.IOException;
+import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
+import java.util.function.Predicate;
 
 import org.jboss.logging.Logger;
 
@@ -59,6 +62,33 @@ public final class ContainerRuntimeUtil {
         }
     }
 
+    private static boolean getRootlessStateFor(ContainerRuntime containerRuntime) {
+        Process rootlessProcess = null;
+        try {
+            ProcessBuilder pb = new ProcessBuilder(containerRuntime.getExecutableName(), "info")
+                    .redirectErrorStream(true);
+            rootlessProcess = pb.start();
+            rootlessProcess.waitFor();
+            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(rootlessProcess.getInputStream()));
+            Predicate<String> stringPredicate;
+            // Docker includes just "rootless" under SecurityOptions, while podman includes "rootless: <boolean>"
+            if (containerRuntime == ContainerRuntime.DOCKER) {
+                stringPredicate = line -> line.trim().equals("rootless");
+            } else {
+                stringPredicate = line -> line.trim().equals("rootless: true");
+            }
+            return bufferedReader.lines().anyMatch(stringPredicate);
+        } catch (IOException | InterruptedException e) {
+            // If an exception is thrown in the process, assume we are not running rootless (default docker installation)
+            log.debugf(e, "Failure to read info output from %s", containerRuntime.getExecutableName());
+            return false;
+        } finally {
+            if (rootlessProcess != null) {
+                rootlessProcess.destroy();
+            }
+        }
+    }
+
     /**
      * Supported Container runtimes
      */
@@ -66,8 +96,18 @@ public final class ContainerRuntimeUtil {
         DOCKER,
         PODMAN;
 
+        private final boolean rootless;
+
+        ContainerRuntime() {
+            this.rootless = getRootlessStateFor(this);
+        }
+
         public String getExecutableName() {
             return this.name().toLowerCase();
+        }
+
+        public boolean isRootless() {
+            return rootless;
         }
     }
 }

